[FLINK-37403] Convert interval join to regular join for updating source

bvarghese1 opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change
- Previously, if inputs to interval join contains updates the planner throws an exception.
- This commit introduces an enhancement to convert interval join to regular join if any of the input to interval join contains updates


## Brief change log

  - Check the input modify kind set for interval join node to see if it contains updates
  - If updates are present convert interval join to regular join node


## Verifying this change

This change added tests and can be verified as follows:

  - Added unit tests for ChangelogModeInference
  - Added restore tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

bvarghese1 committed:
[FLINK-37403] Convert interval join to regular join for updating source

- Previously, if inputs to interval join contains updates the planner throws
  an exception.
- This commit introduces an enhancement to convert interval join to regular join
  if any of the input to interval join contains updates

twalthr commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/optimize/program/FlinkChangelogModeInferenceProgram.scala:
```diff
@@ -287,6 +287,39 @@ class FlinkChangelogModeInferenceProgram extends FlinkOptimizeProgram[StreamOpti
+          ModifyKind.UPDATE) || inputModifyKindSet.contains(ModifyKind.DELETE)
+        if (containsUpdatesOrDeletes) {
+          // Convert to regular join if the input contains Updates
+          val isInnerJoin = intervalJoin.joinSpec.getJoinType == FlinkJoinType.INNER
```
why only inner and not all?

bvarghese1 replied:
This is the same as the RegularJoin condition
RegularJoin has conditions for INNER and SEMI. However, IntervalJoin doesn't support SEMI
Ref: https://github.com/apache/flink/pull/26230/files#diff-ddf95eb3949ab889e8e3bccdacb9f57553aac06657574fd6316ca17dd48321a1R262

twalthr commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/IntervalJoinRestoreTest.java:
```diff
@@ -35,7 +35,13 @@ public IntervalJoinRestoreTest() {
     public List<TableTestProgram> programs() {
         return Arrays.asList(
                 IntervalJoinTestPrograms.INTERVAL_JOIN_EVENT_TIME,
+                IntervalJoinTestPrograms.INTERVAL_JOIN_EVENT_TIME_UPDATING_SOURCE,
```
We don't need a restore test. Restoring doesn't need to be tested. 
I added a `SemanticTestBase` recently to test an execution only:
Take this as an example: https://github.com/apache/flink/pull/26236

twalthr commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/optimize/program/FlinkChangelogModeInferenceProgram.scala:
```diff
@@ -287,6 +287,39 @@ class FlinkChangelogModeInferenceProgram extends FlinkOptimizeProgram[StreamOpti
+            intervalJoin.getCondition,
+            intervalJoin.getJoinType,
+            intervalJoin.getHints)
+          createNewNode(regularJoin, children, providedTrait, requiredTrait, requester)
```
It might be that this approach has one major issue. What is the data type of the timestamp column? Is it still marked as a time attribute or is it a regular column? If it is still marked as a time attribute, windows or subsequent interval joins would still be allowed by the planner. Of course the output would be garbage in this case.

bvarghese1 replied:
Added a commit to materialize time attributes while processing IntervalJoin if the source has updates.

bvarghese1 committed:
[FLINK-37403] Materialize time attributes in IntervalJoin

- Materialize time attributes in IntervalJoin if any of the source contains
  updates
- This will convert IntervalJoin to RegularJoin

davidradl reviewed (commented):
there is a comment on the Jira asking a question; could you address that question please, in case it proves to be a better approach? The question is can we incorporate updates by amending the interval join processing? Changing the interval join to a regular join, seems strange as we are not honouring the requested join type - so could be confusing to the user, would a meaningful error message also be an option. WDYT?
